### DISCUSSION TITLE
ComputeExtraStressConstant - Function input?

### DISCUSSION THREAD

**Discussion by 1runer** [idaholab/moose#21023] [Q&A Modules: Solid mechanics] [answered]
Hi all,

The following question:
We have a prestressed block and are only interested in the possible deformation (and strain/ stress change) until failure happens (e.g. by means of Mohr-Coulomb criterion). 

Now there are various possibilities of initial "prestressing" via PressureBC, Applied Displacements as BC, etc..

I also came across ComputeExtraStressConstant, which however only accepts constant values.

Is there a reason why ComputeExtraStressConstant does not accept a spatial/temporal-dependent function (e.g. from a CSV, postprocessor, z-dependent) as input?

Kind regards

**Comment by GiudGiud** [comment, marked as answer]
Hello

There is a prefactor option in that object so you can use a material property to define the value of the stress
And using a GenericFunctionMaterial you can get a material property that uses a function (spatial and temporal dependent) .
https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html

Guillaume

**Comment by 1runer** [comment]
Hi Guillaume,

thanks for your answer.
I also found this post regarding the usage [https://github.com/idaholab/moose/discussions/17410](url).

But how can I use a function to define this prefactor in at least 3 (or all 6) components independent?
e.g. if I want to add:
stress_zz = z * rho* g, stress_yy =0.7 * stress_zz, stress_xx = 1e6 = const.

How should the 'prefactor' or GenericFunctionMaterial or the ParsedFunction look like to apply this ExtraStress?
I am not quite clear at which point I define the vector or tensor for the prefactor?

Kind regards

**Reply by GiudGiud** [reply]
Hello

ok I missed that you wanted a component dependent prefactor. 
The solution here will be for you to create a new material, which retrieves a tensor material property rather than just a Real material property. Copy paste the ComputeExtraStressConstant.C and ComputeExtraStressConstant.h to a new file name then modify it for a tensor prefactor. 
Please let us know if you need help with that. 

Then once you have that you can use a GenericFunctionRankTwoTensor to define the new prefactor. 
https://mooseframework.inl.gov/source/materials/GenericFunctionRankTwoTensor.html

Guillaume

**Reply by 1runer** [reply]
Hello Guillaume,

thanks for your suggestion, it works fine.

I created my own material and filled the RankTwoTensor of extrastress with some functions from the functionParser.
I have omitted the prefactor, I can add it if needed. 

Thanks a lot and greetings!

**Comment by 1runer** [comment]
Hi Guillaume,

where as the coding worked fine`for me, 
I have another question regarding the solution of the extra stress and the stress update due to ComputeExtraStressConstant material, but also in my material.

When I try to solve a simple problem, the extra_stress is not transferred properly to the stress tensor.
While xx, yy always work fine, zz doesn't work for smaller young's moduli. 
Please see the attached image and the input-file leading to the image.

![extra_stress_only](https://user-images.githubusercontent.com/26140953/169491805-9101435e-ac74-4478-bd60-14430bed5514.PNG)

```[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 10
  ny = 10
  nz = 50
  ymax = 1
  ymin = 0
  xmax = 1
  xmin = 0
  zmin = 0
  zmax = 5
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules]
  [./TensorMechanics]
    [./Master]
      [./all]
      add_variables = true
      strain = SMALL
      incremental = true
      additional_generate_output = 'stress_xx stress_yy stress_zz'
      [../]
    [../]
  [../]
[]

[BCs]
  [./no_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left right'
    value = 0.0
  [../]
  [./no_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom top'
    value = 0.0
  [../]
  [./no_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'back front'
    value = '0.0'
  [../]
[]

[Materials]
  [./Elasticity_tensor]
    type = ComputeElasticityTensor
    fill_method = symmetric_isotropic_E_nu
    C_ijkl = '7.4e7 0'
  [../]
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
  [./extra_stress]
    type = ComputeExtraStressConstant
    extra_stress_tensor = '1 1.5 2 0 0 0'
    prefactor = stress_fac1
    output_properties = 'extra_stress'
    outputs = 'exodus'
  [../]
  [./fac1]
    type = GenericFunctionMaterial
    prop_names = stress_fac1
    prop_values = fac1_fcn
  [../]
[]

[Functions]
  [fac1_fcn]
    type = ParsedFunction
    value = '(z*1e3)'
  []
[]

[Executioner]
  type = Transient
  end_time = 5
  dt = 5
  nl_abs_tol = 1e-10
  l_max_its = 20
  automatic_scaling = true
  compute_scaling_once = true
[]

[Outputs]
  exodus = true
  print_linear_residuals = true
[]
```


Kind regards

**Reply by GiudGiud** [reply]
Hello

~~What is the definition of stress_fac1?~~
I suspect it's just that tensor components arent entered in the right order for zz

How did you implement the tensor prefactor? I still see a function for the prefactor, so a single coefficient. 

Guillaume

**Comment by 1runer** [comment]
Hi Guillaume,

for the example above, I just followed some earlier discussion here on Github [LINK] https://github.com/idaholab/moose/discussions/17410)
The inpute-file above works with the normal TensorMechanics. 

The ExtraStress is calculated using the z-value of each QP.
And as far as I could see, the ExtraStress also agrees with the calculated value.

Only with the transfer into the stress tensor there are problems. Stress_XX, Stress_YY are z-dependent and identical with the ExtraStress, only Stress_ZZ has a constant value for each element. 

Kind regards

**Reply by GiudGiud** [reply]
hello

I see it evolve as expected. 
with this
```
  [./extra_stress]
    type = ComputeExtraStressConstant
    extra_stress_tensor = '1 2 3 4 5 6'
    prefactor = stress_fac1
    output_properties = 'extra_stress'
    outputs = 'exodus'
  [../]
```
I get this output which seems correct?
It has z dependence on every component of the extra_stress, of the expected magnitude. the 22 is the z component. 
![image](https://user-images.githubusercontent.com/25037151/169907238-ac312dd1-7af4-432f-887a-507919fced83.png)

Note that stress_zz is not the extra stress. It s the total stress. And in your case you get strain which diminishes the stress and most importantly here I think it makes the stress more evenly spread.

Guillaume

**Reply by GiudGiud** [reply]
@dschwen if you want to double check on my explanation
